File: F#/chap07/Chap0701.fs
namespace Chap0701

open System

type UnvalidatedCustomerInfo = Undefined
type UnvalidatedAddress = Undefined

type UnvalidatedOrder =
    { OrderId: string
      CustomerInfo: UnvalidatedCustomerInfo
      ShippingAddress: UnvalidatedAddress }

module C070101 =
    type PlaceOrder =
        { OrderForm: UnvalidatedOrder
          Timestamp: DateTime
          UserId: string }

type Command<'data> =
    { Data: 'data
      Timestamp: DateTime
      UserId: string }

type PlaceOrder = Command<UnvalidatedOrder>

type ChangeOrder = Undefined
type CancelOrder = Undefined

type OrderTakingCommand =
    | Place of PlaceOrder
    | Change of ChangeOrder
    | Cancel of CancelOrder

File: Scala/chap07/src/main/scala/Chap0701.scala
import Chap0701.UnvalidatedOrder

import java.time.LocalDateTime

object Chap0701:
  case class UnvalidatedCustomerInfo(value: Any)

  case class UnvalidatedAddress(value: Any)

  case class UnvalidatedOrder(
                               orderId: Int,
                               customerInfo: UnvalidatedCustomerInfo,
                               shippingAddress: UnvalidatedAddress
                             )

  case class Command[A](data: A, timestamp: LocalDateTime, userId: String)

  enum OrderTakingCommand:
    case PlaceOrder(value: Command[UnvalidatedOrder])
    case ChangeOrder(value: Any)
    case CancelOrder(value: Any)

object Chap070101:
  case class PlaceOrder(orderForm: UnvalidatedOrder, timestamp: LocalDateTime, userId: String)
